colcon_core/package_descriptor.py

```diff
@@ -144,11 +138,19 @@ def get_recursive_dependencies(
         return dependencies
 
     def __hash__(self):  # noqa: D105
-        return hash((self.realpath, self.type, self.name))
+        return hash((self.type, self.name))
```

Comment:
Should we use either `(path, type, name)` or just `name` for the hash? Just using two of the three seems arbitrary to me.

Reply:
No. We can't use path, because it can only determine that two paths are equal, not that they are unequal.

Reply:
Why can't two descriptors with the same `realpath` but different `path` have different hashes?

Reply:
Because they compare equal! That's the thing about hashes. If two objects are equal, they *must* have equal hashes. If two objects are unequal, they *likely* (but don't need to) have different hashes. That's the core assumption behind a hash table.

Reply:
Here's a toy example:
```
>>> class Bad:
...   def __hash__(self):
...     return id(self)
...   def __eq__(self, other):
...     return True
...
>>> d = dict()
>>> for i in range(10):
...   d[Bad()]=i
...
>>> print(len(d))
10
```

Reply:
I was thinking the wrong direction altogether. That makes total sense. Sorry for the extra round trip.